### PR TITLE
adds FQDN validation for cluster and for router

### LINKED ISSUE #348: need to validate fqdn (and router fqdn) end with <azure zone>.cloudapp.azure.com

### PR DESCRIPTION
**Pull request by thanasisk**
Fixes #348

### CODE REVIEW

File: pkg/validate/validate.go [hunk 1]
```diff
@@ -29,6 +29,16 @@ var validAgentPoolProfileNames = map[string]struct{}{
 	string(api.AgentPoolProfileRoleMaster):  struct{}{},
 }
 
+// az account list-locations will fetch the latest additions, if needed
+var azureZones = regexp.MustCompile(`{` +
+	`eastasia|southeastasia|centralus|eastus|eastus2|westus|northcentralus|` +
+	`southcentralus|northeurope|westeurope|japanwest|japaneast|brazilsouth|` +
+	`australiaeast|australiasoutheast|southindia|centralindia|westindia|` +
```
Comment (jim-minter) [inline pkg/validate/validate.go]: I don't think we want to hard-code these ourselves and I'm not aware of a canonical location to pull these from.  Think it'll have to be something like `\w+\.cloudapp\.azure\.com`

Comment (charlesakalugwu) [reply pkg/validate/validate.go]: looks like there's an api to retrieve these from
`
az account list-locations | jq .[].name --raw-output
`

Comment (thanasisk) [reply pkg/validate/validate.go]: @charlesakalugwu true but how can we fetch it dynamically into the code without
1) requiring always on-line
2) requiring az credentials

I guess I will follow Jim's suggestion, which was my initial solution

**Comment by jim-minter** [comment]
rebase onto #459 and this will be much simpler

**Comment by kargakis** [comment]
/hold

can you add a unit test that uses an invalid hostname?

File: pkg/util/fixtures/fixtures.go [hunk 2]
```diff
@@ -10,12 +10,12 @@ func NewTestOpenShiftCluster() *api.OpenShiftManagedCluster {
 		Name:     "openshift",
 		Config:   &api.Config{},
 		Properties: &api.Properties{
-			FQDN:             "www.example.com",
+			FQDN:             "www.example.eastus.cloudapp.azure.com",
```
Comment (jim-minter) [inline pkg/util/fixtures/fixtures.go]: will necessarily be foo.region.cloudapp.azure.com - can validate the number of domain parts

File: pkg/validate/validate.go [hunk 3]
```diff
@@ -41,6 +43,10 @@ func isValidHostname(h string) bool {
 	return len(h) <= 255 && rxRfc1123.MatchString(h)
 }
 
+func isAzureZone(h string) bool {
+	return azureZones.MatchString(h)
```
Comment (jim-minter) [inline pkg/validate/validate.go]: use strings.HasSuffix() and also count the number of domain parts

File: pkg/validate/validate_test.go [hunk 4]
```diff
@@ -314,3 +314,17 @@ func TestValidate(t *testing.T) {
 		}
 	}
 }
+
+func TestInvalidAzureZone(t *testing.T) {
```
Comment (jim-minter) [inline pkg/validate/validate_test.go]: Condense to one function, TestIsAzureZone

Comment (kargakis) [inline pkg/validate/validate_test.go]: I imagined you'd be able to add a test case in `TestValidate`.

File: pkg/validate/validate.go [hunk 5]
```diff
@@ -41,6 +42,10 @@ func isValidHostname(h string) bool {
 	return len(h) <= 255 && rxRfc1123.MatchString(h)
 }
 
+func isAzureZone(fqdn string) bool {
+	return strings.HasSuffix(fqdn, "cloudapp.azure.com") && len(strings.Split(fqdn, ".")) == 5
```
Comment (jim-minter) [inline pkg/validate/validate.go]: nits: need a preceding `.` on `".cloudapp.azure.com"`, can use strings.Count()

File: pkg/validate/validate_test.go [hunk 6]
```diff
@@ -314,3 +314,14 @@ func TestValidate(t *testing.T) {
 		}
 	}
 }
+
+func TestIsAzureZone(t *testing.T) {
```
Comment (jim-minter) [inline pkg/validate/validate_test.go]: add test for `"too.long.domain.cloudapp.azure.com"`

**Comment by jim-minter** [comment]
lgtm, please squash the commits and will merge

**Pull request merged**